### ISSUE TITLE
implicit-clone: PartialEq not impl'd for &Ty

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#10019]
### Summary

.

### Lint Name

implicit-clone

### Reproducer

I tried this code:

```rust
fn main() {
    let x = &Baz;
    let y = &Baz;

    y.to_owned() == *x;
}

#[derive(PartialEq)]
struct Baz;

impl ToOwned for Baz {
    type Owned = Baz;
    fn to_owned(&self) -> Baz {
        Baz
    }
}
```

I saw this happen:

```
This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0277]: can't compare `&Baz` with `Baz`
 --> src/main.rs:5:23
  |
5 |     let _ = y.clone() == *x;
  |                       ^^ no implementation for `&Baz == Baz`
  |
  = help: the trait `std::cmp::PartialEq<Baz>` is not implemented for `&Baz`
  = help: the trait `std::cmp::PartialEq` is implemented for `Baz`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
Original diagnostics will follow.

warning: implicitly cloning a `Baz` by calling `to_owned` on its dereferenced type
 --> src/main.rs:5:5
  |
5 |     y.to_owned() == *x;
  |     ^^^^^^^^^^^^ help: consider using: `y.clone()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#implicit_clone
  = note: requested on the command line with `-W clippy::implicit-clone`
```

### Version

```text
rustc 1.67.0-nightly (c97b539e4 2022-11-30)
binary: rustc
commit-hash: c97b539e408ea353f4fde2f9251d598291fec421
commit-date: 2022-11-30
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
